lib/fs/walkfs.go

```diff
@@ -10,7 +10,55 @@
 
 package fs
 
-import "path/filepath"
+import (
+	"fmt"
+	"path/filepath"
+)
+
+func newAncestorDirList() *ancestorDirList {
+	l.Debugf("ancestorDirList: new")
+	return &ancestorDirList{list: make([]FileInfo, 0, 20)}
+}
+
+type ancestorDirList struct {
+	list []FileInfo
+}
+
+func (ancestors *ancestorDirList) AppendUnlessPresent(info FileInfo) bool {
+	l.Debugf("ancestorDirList: AppendUnlessPresent '%s'", info.Name())
+	for _, ancestor := range ancestors.list {
+		if SameFile(info, ancestor) {
+			return false
+		}
+	}
+	aLen := len(ancestors.list)
+	if aLen == cap(ancestors.list) {
+		t := make([]FileInfo, aLen, aLen*2)
+		copy(t, ancestors.list)
+		ancestors.list = t
+		l.Debugf("ancestorDirList: capacity increased to %d", cap(ancestors.list))
+	}
+	ancestors.list = append(ancestors.list, info)
+	return true
+}
+
+func (ancestors *ancestorDirList) Remove(info FileInfo) bool {
```

Comment:
I think I'd call this `Pop() FileInfo` and do the check/panic on the other side maybe (or not at all, as it would require recursion not to be recursion or something like that to actually trigger)?

With the `Identity()` stuff above that could be used here as well as it can't fail, if we want to be paranoid about it.

Reply:
As I answered above, I preferred not to use the identity, because I would have to program that myself, even though they already do that in os.fileStat. The only available function to access the identity stuff that is already inside the FileInfo objects is os.SameFile. Hence I used it...

Reply:
On the other side I mean the call site, i.e. essentially:

```
ancestors.Push(info)
defer func() {
  popped := ancestors.Pop()
  if !SameFile(info, popped) {
    panic("wat")
  }
}()
```

that also highlights further what would be the (impossible?) scenario where a panic would be warranted.

Reply:
I wanted to hide the usage of SameFile inside the ancestors implementation. I can change that, if you want, no problem. 

In fact, I conceived ancestors as a stack or set of unique elements. The name Push of the method wouldn't suggest that it may fail to insert the element, hence I could call it PushUnlessPresent. The name Remove was from the idea of set rather than a stack, I'll rename it.

Reply:
fixed